### ISSUE TITLE
Add capabilities from other roles

### ISSUE THREAD

**Issue by triplem** [debops/ansible-postfix#18]
Is it possible to add a capability from another role? eg. on dovecot we would need the 'auth' capability (which is not documented and most probabyl not official right now).

**Comment by drybjed** [comment]
TL;DR: this is possible, but if you do this you lose ansible idempotency.

Postfix capabilities are designed to be managed from Ansible inventory. Basically, you can decide what is the purpose of a particular host - is it an MX gateway, or does it have local/virtual mail, are all mails archived somewhere, etc. When you do that from Ansible inventory, everything is fine all the time. Let's call a particular set of capabilities `alpha`.

But when you start messing with Postfix capabilities from another role using role dependency variables, Postfix role assumes different set of capabilities than before, say `beta`. And this set is active only during the execution of that particular role. After that, if Postfix role happens to be called again, it will use either another set of capabilities, `theta`, or capabilities defined in inventory, `alpha`. Which makes capabilities from `beta` invalidated.

On the next Ansible run, Postfix again will start with capabilities `alpha`, which makes capabilities `beta` and `theta` invalidated. Then, it executes your role that sets capabilities `theta` and the same cycle is repeated again.

So, as you can see - you could do that, but your settings won't be applied forever.

Instead what you would want to do in your role (and what I've done in `debops.mailman` is to check what Postfix capabilities are enabled, using `ansible_local.postfix.capabilities` fact (if it's defined) and act accordingly - if a particular capability is not present, don't try to use it, etc. You could then write in the readme that certain Postfix capabilities are recommended for full functionality, but let the user decide which ones are active on a host.

**Comment by triplem** [comment]
I see, I have now added the same parameters as already defined in the "sasl_auth.j2"-template, which I wanted to avoid. But you do have a point there, and I do like to follow this best-practices.

Another question though. Dovecot needs some settings in the smtpd_recipient_restrictions. and I don't like to overwrite already existing content in there. Could I add content to this variable in another role without loosing content set by postfix or even other roles?

As soon as I am finished I will come up with a dovecot role, which provides some dovecot things. Hope that you are interested.

**Comment by drybjed** [comment]
Postfix role exposes a mechanism to add your own configuration to existing variables, via Ansible local facts. You can see it in `debops.mailman` and `debops.smstools` roles `meta/main.yml` files. Using it, you can add elements to various Postfix parameters or even add your own configuration variables. Removal is currently not supported, you can only append your settings.

However, `smtpd_*_restrictions` configuration in Postfix is an unique case, because the order of entries in it is important. I'm not sure how to expose it for other roles in a clean way, at the moment I have an idea about creating a set of "weighted hook lists" numbered from 0 to 9, to which you could add your own entries and in almost all cases they could be correctly sorted, but that's not yet implemented. For now, if you want to mess with `smtpd_*_restrictions`, I would suggest just redefining them via inventory (definition in inventory will overwrite the dynamically generated version, so you need to redefine them from scratch).

I'm most certainly interested in a Dovecot role. :-) However, I wouldn't necessarily add a dependency on Postfix to it. Instead, I would configure Postfix and Dovecot independently (I think Dovecot needs a similar "capabilities" functionality to be flexible), and use a third role to configure your mail stack, which would use Postfix and Dovecot roles as dependencies. That way, you could have a role that configures Postfixadmin, or a role that configures `vmm`, or a role with custom mail authentication stack. Think about it like a webapp stack - you have MySQl and nginx roles that configure services, and then you have phpIPAM role that uses these roles as dependencies to configure what the application needs.

**Comment by drybjed** [comment]
The redesigned `debops.postfix` role allows for creation of custom Ansible roles that add their own configuration or modify existing one, for example [`debops.postconf`](https://github.com/debops/ansible-postconf/). i'm closing this issue as resolved.
